### ISSUE TITLE
Add containerd support to "--local" mode

### ISSUE THREAD

**Issue by antoineco** [google/ko#147]
Running `ko` with `--local` is super convenient 💯  Unfortunately it seems to works only with Docker daemons (correct me if I'm wrong).

I recently switched to `k3s` in my local setup, which does not rely on a Docker daemon. Instead, `k3s` exposes a `containerd` GRPC server at `/run/k3s/containerd/containerd.sock`.

What do you think about supporting publishing images directly to `containerd`, and enabling the feature via and environment variable, e.g. `KO_CONTAINERD_HOST`?

**Comment by antoineco** [comment]
Ref #149

**Comment by jonjohnsonjr** [comment]
I'm certainly open to it.

**Comment by antoineco** [comment]
This issue has suddenly become [very relevant](https://kubernetes.io/blog/2020/12/02/dont-panic-kubernetes-and-docker/) :)

**Comment by jonjohnsonjr** [comment]
Is this as simple as [`containerd.New`](https://pkg.go.dev/github.com/containerd/containerd?readme=expanded#New) and then [`client.Import`](https://pkg.go.dev/github.com/containerd/containerd?readme=expanded#Client.Import)? That shouldn't be too hard to add a new publisher and wire it up with flags.

**Comment by ImJasonH** [comment]
/reopen

Still relevant AFAIK

**Comment by leoluk** [comment]
A simple (albeit inefficient) workaround is to export a tarball using `--tarball=foo.tar --push=false` and import the resulting image into k3s/containerd using `k3s ctr images import`.

Simple example usage: https://github.com/monogon-dev/NetMeta/blob/94fcccddcb03faef99ce46ea4738b5d14804ffb8/scripts/build_containers.sh#L17

**Comment by jonjohnsonjr** [comment]
A more efficient strategy might be to just run a very simple registry implementation in-process and have `ko` act as the registry? Serializing to tarballs is annoying because you lose deduplication...

**Comment by leoluk** [comment]
> A more efficient strategy might be to just run a very simple registry implementation in-process and have `ko` act as the registry? Serializing to tarballs is annoying because you lose deduplication...

It should be possible to just write the layers straight to the containerd local store, right? Just like a Docker build does.

**Comment by jonjohnsonjr** [comment]
> It should be possible to just write the layers straight to the containerd local store, right?

Definitely, but IMO it would be really nice if we could avoid it. I don't _really_ want to link the containerd client library.

OTOH, there are cases where containerd might be on another machine, and my idea doesn't work if the target runtime is on a different network.
